### PR TITLE
Introduce WebAuthnManager and deprecate existing ContextValidator(s)

### PR DESCRIPTION
**Pull request by ynojima**
As existing `ContextValidator(s)` does data parse and validation in the same method call ( `ContextValidator#validate` ), caller cannot access the parsed data when an error occurred in the validation phase. 
To resolve this issue, introduce `WebAuthnManager` class which splits data parse phase and data validation phase into two dedicated method call.

Sample code:
```java

WebAuthnRegistrationRequest webAuthnRegistrationRequest = new WebAuthnRegistrationRequest(attestationObject, clientDataJSON, clientExtensionJSON, transports);
WebAuthnRegistrationParameters webAuthnRegistrationParameters = new WebAuthnRegistrationParameters(serverProperty, userVerificationRequired, userPresenceRequired, expectedExtensionIds);
WebAuthnRegistrationData webAuthnRegistrationData;
try{
    // data parse phase
    webAuthnRegistrationData = webAuthnManager.parseRegistrationRequest(webAuthnRegistrationRequest);
}
catch (DataConversionException e){
    // If you would like to handle WebAuthn data structure parse error, please catch DataConversionException
    throw e;
}
try{
    // data validation phase
    webAuthnRegistrationData.validate(webAuthnRegistrationParameters);
}
catch (ValidationException e){
    // If you would like to handle WebAuthn data validation error, please catch ValidationException
    throw e;
}
```

`WebAuthnManager` deprecates existing `WebAuthnRegistrationContextValidator` and `WebAuthnAuthenticationContextValidator`. These classes will be removed by 1.0 GA.

### CODE REVIEW

File: webauthn4j-core/src/test/java/com/webauthn4j/data/WebAuthnAuthenticationDataTest.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2002-2018 the original author or authors.
```
Comment (tnorimat) [inline webauthn4j-core/src/test/java/com/webauthn4j/data/WebAuthnAuthenticationDataTest.java]: IMO, "Copyright 2002-2019" might be better.

File: webauthn4j-core/src/test/java/com/webauthn4j/data/WebAuthnAuthenticationParametersTest.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2002-2018 the original author or authors.
```
Comment (tnorimat) [inline webauthn4j-core/src/test/java/com/webauthn4j/data/WebAuthnAuthenticationParametersTest.java]: IMO, "Copyright 2002-2019" might be better.

File: webauthn4j-core/src/test/java/com/webauthn4j/data/WebAuthnAuthenticationRequestTest.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2002-2018 the original author or authors.
```
Comment (tnorimat) [inline webauthn4j-core/src/test/java/com/webauthn4j/data/WebAuthnAuthenticationRequestTest.java]: IMO, "Copyright 2002-2019" might be better.

File: webauthn4j-core/src/test/java/com/webauthn4j/data/WebAuthnRegistrationDataTest.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2002-2018 the original author or authors.
```
Comment (tnorimat) [inline webauthn4j-core/src/test/java/com/webauthn4j/data/WebAuthnRegistrationDataTest.java]: IMO, "Copyright 2002-2019" might be better.

File: webauthn4j-core/src/test/java/com/webauthn4j/data/WebAuthnRegistrationParametersTest.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2002-2018 the original author or authors.
```
Comment (tnorimat) [inline webauthn4j-core/src/test/java/com/webauthn4j/data/WebAuthnRegistrationParametersTest.java]: IMO, "Copyright 2002-2019" might be better.

File: webauthn4j-core/src/test/java/com/webauthn4j/data/WebAuthnRegistrationRequestTest.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright 2002-2018 the original author or authors.
```
Comment (tnorimat) [inline webauthn4j-core/src/test/java/com/webauthn4j/data/WebAuthnRegistrationRequestTest.java]: IMO, "Copyright 2002-2019" might be better.

File: webauthn4j-core/src/test/java/com/webauthn4j/validator/BeanAssertUtilTest.java [hunk 7]
```diff
@@ -143,8 +143,9 @@ void validate_WebAuthnRegistrationContext_test() {
```
Comment (tnorimat) [inline webauthn4j-core/src/test/java/com/webauthn4j/validator/BeanAssertUtilTest.java]: IMO, it seems to be better to write tests for the newly introduced methods :
``` java
validate(WebAuthnRegistrationData)
validate(WebAuthnRegistrationParameters)
validate(WebAuthnAuthenticationData)
validate(WebAuthnAuthenticationParameters)
```

File: webauthn4j-core/src/test/java/com/webauthn4j/validator/WebAuthnAuthenticationDataValidatorTest.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2002-2018 the original author or authors.
```
Comment (tnorimat) [inline webauthn4j-core/src/test/java/com/webauthn4j/validator/WebAuthnAuthenticationDataValidatorTest.java]: IMO, "Copyright 2002-2019" might be better.

File: webauthn4j-core/src/test/java/com/webauthn4j/validator/WebAuthnRegistrationDataValidatorTest.java [hunk 9]
```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2002-2018 the original author or authors.
```
Comment (tnorimat) [inline webauthn4j-core/src/test/java/com/webauthn4j/validator/WebAuthnRegistrationDataValidatorTest.java]: IMO, "Copyright 2002-2019" might be better.

File: webauthn4j-core/src/test/java/sample/WebAuthnManagerSample.java [hunk 10]
```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright 2002-2018 the original author or authors.
```
Comment (tnorimat) [inline webauthn4j-core/src/test/java/sample/WebAuthnManagerSample.java]: IMO, "Copyright 2002-2019" might be better.

File: webauthn4j-core/src/main/java/com/webauthn4j/WebAuthnManager.java [hunk 11]
```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright 2002-2018 the original author or authors.
```
[the remark this answers is not in the document]
Comment (ynojima) [reply webauthn4j-core/src/main/java/com/webauthn4j/WebAuthnManager.java]: I'd like to address this issue in a dedicated pull-request as it is not limited to this pull-request.

File: webauthn4j-core/src/main/java/com/webauthn4j/validator/WebAuthnAuthenticationDataValidator.java [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+        List<String> expectedExtensionIdentifiers = webAuthnAuthenticationParameters.getExpectedExtensionIds();
+        extensionValidator.validate(clientExtensions, authenticationExtensionsAuthenticatorOutputs, expectedExtensionIdentifiers);
+
+        //spec| Using the credential public key, validate that sig is a valid signature over
+        //spec| the binary concatenation of the authenticatorData and the hash of the collectedClientData.
+        assertionSignatureValidator.validate(webAuthnAuthenticationData, authenticator.getAttestedCredentialData().getCOSEKey());
+
+        //spec| Step17
+        //spec| If the signature counter value adata.signCount is nonzero or the value stored in conjunction with
+        //spec| credential’s id attribute is nonzero, then run the following sub-step:
```
[the remark this answers is not in the document]
Comment (ynojima) [reply webauthn4j-core/src/main/java/com/webauthn4j/validator/WebAuthnAuthenticationDataValidator.java]: As this is copied from the spec, I'd like to leave as it is.

File: webauthn4j-core/src/main/java/com/webauthn4j/validator/WebAuthnAuthenticationDataValidator.java [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+        long presentedCounter = authenticatorData.getSignCount();
+        long storedCounter = authenticator.getCounter();
+        if (presentedCounter > 0 || storedCounter > 0) {
+            //spec| If the signature counter value adata.signCount is
+            //spec| greater than the signature counter value stored in conjunction with credential’s id attribute.
+            if (presentedCounter > storedCounter) {
+
+                //spec| Update the stored signature counter value, associated with credential’s id attribute, to be the value of authData.signCount.
```
[the remark this answers is not in the document]
Comment (ynojima) [reply webauthn4j-core/src/main/java/com/webauthn4j/validator/WebAuthnAuthenticationDataValidator.java]: As this is copied from the spec, I'd like to leave as it is.

File: webauthn4j-core/src/main/java/com/webauthn4j/validator/WebAuthnAuthenticationDataValidator.java [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+
+                //      (caller need to update the signature counter value based on the value set in the Authenticator instance)
+                authenticator.setCounter(presentedCounter);
+            }
+            //spec| less than or equal to the signature counter value stored in conjunction with credential’s id attribute.
```
[the remark this answers is not in the document]
Comment (ynojima) [reply webauthn4j-core/src/main/java/com/webauthn4j/validator/WebAuthnAuthenticationDataValidator.java]: As this is copied from the spec, I'd like to leave as it is.

File: webauthn4j-core/src/main/java/com/webauthn4j/validator/WebAuthnRegistrationContextValidationResponse.java [hunk 15]
```diff
@@ -25,7 +26,9 @@
 
 /**
  * Envelope class for WebAuthnRegistrationContext validation result
+ * @deprecated {@link WebAuthnRegistrationContextValidator} is deprecated. please use {@link WebAuthnManager} instead.
```
[the remark this answers is not in the document]
Comment (ynojima) [reply webauthn4j-core/src/main/java/com/webauthn4j/validator/WebAuthnRegistrationContextValidationResponse.java]: Correct.

File: webauthn4j-core/src/main/java/com/webauthn4j/validator/WebAuthnRegistrationDataValidator.java [hunk 16]
```diff
@@ -0,0 +1,219 @@
...
+        //spec| Check that the credentialId is not yet registered to any other user. If registration is requested for
+        //spec| a credential that is already registered to a different user, the Relying Party SHOULD fail this registration
+        //spec| ceremony, or it MAY decide to accept the registration, e.g. while deleting the older registration.
+
+        //      (This step is out of WebAuthn4J scope. It's caller's responsibility.)
+
+        //spec| Step18
+        //spec| If the attestation statement attStmt verified successfully and is found to be trustworthy,
+        //spec| then register the new credential with the account that was denoted in the options.user passed to create(),
+        //spec| by associating it with the credential ID and credential public key contained in authData’s attestation data,
```
[the remark this answers is not in the document]
Comment (ynojima) [reply webauthn4j-core/src/main/java/com/webauthn4j/validator/WebAuthnRegistrationDataValidator.java]: As this is copied from the spec, I'd like to leave as it is.

File: webauthn4j-core/src/main/java/com/webauthn4j/validator/WebAuthnAuthenticationDataValidator.java [hunk 17]
```diff
@@ -0,0 +1,224 @@
...
+        //spec| verify that credential.response.userHandle is present, and that the user identified by this value is the owner of credentialSource.
+        //      (This step is out of WebAuthn4J scope. It's caller's responsibility.)
+
+        //spec| Step3
+        //spec| Using credential’s id attribute (or the corresponding rawId, if base64url encoding is inappropriate for your use case),
+        //spec| look up the corresponding credential public key.
+        //      (This step is out of WebAuthn4J scope. It's caller's responsibility.)
+
+        //spec| Step4
+        //spec| Let cData, aData and sig denote the value of credential’s response's clientDataJSON, authenticatorData,
```
[the remark this answers is not in the document]
Comment (ynojima) [reply webauthn4j-core/src/main/java/com/webauthn4j/validator/WebAuthnAuthenticationDataValidator.java]: As this is copied from the spec, I'd like to leave as it is.

File: webauthn4j-core/src/main/java/com/webauthn4j/validator/WebAuthnAuthenticationDataValidator.java [hunk 18]
```diff
@@ -0,0 +1,224 @@
...
+        //spec| If the user was identified before the authentication ceremony was initiated,
+        //spec| verify that the identified user is the owner of credentialSource.
+        //spec| If credential.response.userHandle is present,
+        //spec| verify that this value identifies the same user as was previously identified.
+        //spec| If the user was not identified before the authentication ceremony was initiated,
+        //spec| verify that credential.response.userHandle is present, and that the user identified by this value is the owner of credentialSource.
+        //      (This step is out of WebAuthn4J scope. It's caller's responsibility.)
+
+        //spec| Step3
+        //spec| Using credential’s id attribute (or the corresponding rawId, if base64url encoding is inappropriate for your use case),
```
[the remark this answers is not in the document]
Comment (ynojima) [reply webauthn4j-core/src/main/java/com/webauthn4j/validator/WebAuthnAuthenticationDataValidator.java]: As this is copied from the spec, I'd like to leave as it is.

File: webauthn4j-core/src/main/java/com/webauthn4j/validator/WebAuthnAuthenticationContextValidationResponse.java [hunk 19]
```diff
@@ -25,7 +26,9 @@
 
 /**
  * Envelope class for WebAuthnAuthenticationContext validation result
+ * @deprecated {@link WebAuthnAuthenticationContextValidator} is deprecated. please use {@link WebAuthnManager} instead.
```
[the remark this answers is not in the document]
Comment (ynojima) [reply webauthn4j-core/src/main/java/com/webauthn4j/validator/WebAuthnAuthenticationContextValidationResponse.java]: Correct.

File: webauthn4j-core/src/main/java/com/webauthn4j/WebAuthnManager.java [hunk 20]
```diff
@@ -0,0 +1,197 @@
...
+
+
+    public WebAuthnRegistrationData parseRegistrationRequest(WebAuthnRegistrationRequest webAuthnRegistrationRequest) throws DataConversionException {
+
+        byte[] clientDataBytes = webAuthnRegistrationRequest.getClientDataJSON();
+        byte[] attestationObjectBytes = webAuthnRegistrationRequest.getAttestationObject();
+
+        CollectedClientData collectedClientData = collectedClientDataConverter.convert(clientDataBytes);
+        AttestationObject attestationObject = attestationObjectConverter.convert(attestationObjectBytes);
+        Set<AuthenticatorTransport> transports = authenticatorTransportConverter.convertSet(webAuthnRegistrationRequest.getTransports());
```
[the remark this answers is not in the document]
Comment (ynojima) [reply webauthn4j-core/src/main/java/com/webauthn4j/WebAuthnManager.java]: At this point, yes. `AuthenticatorTransport` is not needed on parsing and validating the public-key. That's why `WebAuthnRegistrationRequest` has a constructor without `AuthenticatorTransport`.

File: webauthn4j-core/src/main/java/com/webauthn4j/data/WebAuthnAuthenticationParameters.java [hunk 21]
```diff
@@ -0,0 +1,136 @@
...
+import java.util.Objects;
+
+public class WebAuthnAuthenticationParameters {
+
+    private final ServerProperty serverProperty;
+    private final Authenticator authenticator;
+
+    // verification condition
+    private final LocalDateTime timestamp;
+    private boolean userVerificationRequired;
```
[the remark this answers is not in the document]
Comment (ynojima) [reply webauthn4j-core/src/main/java/com/webauthn4j/data/WebAuthnAuthenticationParameters.java]: This is an unintended left. I've removed.

File: webauthn4j-core/src/main/java/com/webauthn4j/WebAuthnManager.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (tnorimat) [reply webauthn4j-core/src/main/java/com/webauthn4j/WebAuthnManager.java]: I see.

File: webauthn4j-core/src/main/java/com/webauthn4j/WebAuthnManager.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (tnorimat) [reply webauthn4j-core/src/main/java/com/webauthn4j/WebAuthnManager.java]: I see.

File: webauthn4j-core/src/main/java/com/webauthn4j/data/WebAuthnAuthenticationParameters.java (hunk 21 above)
[the remark this answers is not in the document]
Comment (tnorimat) [reply webauthn4j-core/src/main/java/com/webauthn4j/data/WebAuthnAuthenticationParameters.java]: I see.

File: webauthn4j-core/src/main/java/com/webauthn4j/validator/WebAuthnAuthenticationContextValidationResponse.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (tnorimat) [reply webauthn4j-core/src/main/java/com/webauthn4j/validator/WebAuthnAuthenticationContextValidationResponse.java]: Thanks.

File: webauthn4j-core/src/main/java/com/webauthn4j/validator/WebAuthnAuthenticationDataValidator.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (tnorimat) [reply webauthn4j-core/src/main/java/com/webauthn4j/validator/WebAuthnAuthenticationDataValidator.java]: I see.

File: webauthn4j-core/src/main/java/com/webauthn4j/validator/WebAuthnAuthenticationDataValidator.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (tnorimat) [reply webauthn4j-core/src/main/java/com/webauthn4j/validator/WebAuthnAuthenticationDataValidator.java]: I see.

File: webauthn4j-core/src/main/java/com/webauthn4j/validator/WebAuthnAuthenticationDataValidator.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (tnorimat) [reply webauthn4j-core/src/main/java/com/webauthn4j/validator/WebAuthnAuthenticationDataValidator.java]: I see.

File: webauthn4j-core/src/main/java/com/webauthn4j/validator/WebAuthnAuthenticationDataValidator.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (tnorimat) [reply webauthn4j-core/src/main/java/com/webauthn4j/validator/WebAuthnAuthenticationDataValidator.java]: I see.

File: webauthn4j-core/src/main/java/com/webauthn4j/validator/WebAuthnAuthenticationDataValidator.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (tnorimat) [reply webauthn4j-core/src/main/java/com/webauthn4j/validator/WebAuthnAuthenticationDataValidator.java]: I see.

File: webauthn4j-core/src/main/java/com/webauthn4j/validator/WebAuthnRegistrationContextValidationResponse.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (tnorimat) [reply webauthn4j-core/src/main/java/com/webauthn4j/validator/WebAuthnRegistrationContextValidationResponse.java]: Thanks.

File: webauthn4j-core/src/main/java/com/webauthn4j/validator/WebAuthnRegistrationDataValidator.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (tnorimat) [reply webauthn4j-core/src/main/java/com/webauthn4j/validator/WebAuthnRegistrationDataValidator.java]: I see.

**Comment by tnorimat** [comment]
The following unit test error raised :
com.webauthn4j.validator.BeanAssertUtilTest > validate_WebAuthnRegistrationParameters_test() FAILED
```java
    @Test
    void validate_WebAuthnRegistrationParameters_test(){
        WebAuthnRegistrationParameters webAuthnRegistrationParameters = new WebAuthnRegistrationParameters(
                mock(ServerProperty.class),
                true
        );
        BeanAssertUtil.validate(webAuthnRegistrationParameters);
    }
```
Using `mock(ServerProperty.class)`, this `.getRpId()` returns null which caused the error just above.

IMO, how about instantiating `ServerProperty` instead using `mock()` ?

File: webauthn4j-core/src/test/java/com/webauthn4j/validator/BeanAssertUtilTest.java (hunk 7 above)
Comment (ynojima) [reply webauthn4j-core/src/test/java/com/webauthn4j/validator/BeanAssertUtilTest.java]: Thank you for pointing out. I add test cases in https://github.com/webauthn4j/webauthn4j/pull/183/commits/cd318f6053d389b2328d79369d842db4a02b23bd

**Comment by ynojima** [comment]
> The following unit test error raised :
> com.webauthn4j.validator.BeanAssertUtilTest > validate_WebAuthnRegistrationParameters_test() FAILED
> 
> ```java
>     @Test
>     void validate_WebAuthnRegistrationParameters_test(){
>         WebAuthnRegistrationParameters webAuthnRegistrationParameters = new WebAuthnRegistrationParameters(
>                 mock(ServerProperty.class),
>                 true
>         );
>         BeanAssertUtil.validate(webAuthnRegistrationParameters);
>     }
> ```
> 
> Using `mock(ServerProperty.class)`, this `.getRpId()` returns null which caused the error just above.
> 
> IMO, how about instantiating `ServerProperty` instead using `mock()` ?

Fixed in https://github.com/webauthn4j/webauthn4j/pull/183/commits/c2f3b27bf9f36d277e682416453c16fb3fe3dfd8

**Pull request merged**